frontend/server/app.test.ts

```diff
@@ -626,23 +626,314 @@ describe('UIServer apis', () => {
     });
   });
 
-  // TODO: Add integration tests for k8s helper related endpoints
+  describe('/apps/tensorboard', () => {
+    let request: requests.SuperTest<requests.Test>;
+    let k8sGetCustomObjectSpy: jest.SpyInstance;
+    let k8sDeleteCustomObjectSpy: jest.SpyInstance;
+    let k8sCreateCustomObjectSpy: jest.SpyInstance;
+    function newGetTensorboardResponse({
+      name = 'viewer-example',
+      logDir = 'log-dir-example',
+      tensorflowImage = 'tensorflow:2.0.0',
+      type = 'tensorboard',
+    }: {
+      name?: string;
+      logDir?: string;
+      tensorflowImage?: string;
+      type?: string;
+    } = {}) {
+      return {
+        response: undefined as any, // unused
+        body: {
+          metadata: {
+            name,
+          },
+          spec: {
+            tensorboardSpec: { logDir, tensorflowImage },
+            type,
+          },
+        },
+      };
+    }
+    beforeEach(() => {
+      app = new UIServer(loadConfigs(argv, {}));
+      request = requests(app.start());
+      k8sGetCustomObjectSpy = jest.spyOn(
+        K8S_TEST_EXPORT.k8sV1CustomObjectClient,
+        'getNamespacedCustomObject',
+      );
+      k8sDeleteCustomObjectSpy = jest.spyOn(
+        K8S_TEST_EXPORT.k8sV1CustomObjectClient,
+        'deleteNamespacedCustomObject',
+      );
+      k8sCreateCustomObjectSpy = jest.spyOn(
+        K8S_TEST_EXPORT.k8sV1CustomObjectClient,
+        'createNamespacedCustomObject',
+      );
+    });
+
+    describe('get', () => {
+      it('requires logdir for get tensorboard', done => {
+        request.get('/apps/tensorboard').expect(404, 'logdir argument is required', done);
+      });
+
+      it('requires namespace for get tensorboard', done => {
+        request
+          .get('/apps/tensorboard?logdir=some-log-dir')
+          .expect(404, 'namespace argument is required', done);
+      });
+
+      it('does not crash with a weird query', done => {
+        k8sGetCustomObjectSpy.mockImplementation(() =>
+          Promise.resolve(newGetTensorboardResponse()),
+        );
+        // The special case is that, decodeURIComponent('%2') throws an
+        // exception, so this can verify handler doesn't do extra
+        // decodeURIComponent on queries.
+        const weirdLogDir = encodeURIComponent('%2');
+        request.get(`/apps/tensorboard?logdir=${weirdLogDir}&namespace=test-ns`).expect(200, done);
+      });
```

Comment:
@eterna2 I think the confusion you found with `encodeURIComponent` and `decodeURIComponent` can be demonstrated by this unit test.

This was initially failing with decodeURIComponent throwing on invalid code '%2'. After I removed all the `decodeURIComponent` calls in node server, it's passing. Therefore, the problem was that express automatically parses `req.query.xxx` for us, so we don't need to decode it by ourselves each time. However, most node server handlers decoded for an extra time, so our UI had to encode an extra time too to let these work together.

Now, the fix is clear, we should let node server not call `decodeURIComponent` manually and let UI only encode for once.

Reply:
Do u want me to raise a PR to fix this? Or r u fixing it?

Reply:
Thanks! I've fixed it for tensorboard handlers.
I will create an issue to track related fixes, just FYI to keep you in the loop.

Reply:
And if you'd like to provide help, that will definitely welcome!

Reply:
Created issue: https://github.com/kubeflow/pipelines/issues/3362